v2_ex46.py

amymhaddad opened the pull request:
I use csv for this exercise.

amymhaddad committed:
v2_ex46.py

I use csv for this exercise.

amymhaddad commented on v2_ex46.py:
```diff
@@ -0,0 +1,30 @@
+    if word in string_of_words:
+        word_count = string_of_words.count(word)
+        final_output += f"{word}: {word_count}\n"
+print(final_output.strip())
```
@paulghaddad Question here: I'm having trouble getting this to print highest to lowest (ie, badger: 7 should appear first, b/c it's the most frequently used). Any suggestions?

paulghaddad commented on v2_ex46.py:
```diff
@@ -0,0 +1,30 @@
+"""
+
+import csv
+new_words = [words for words in csv.reader(open("text_file.txt", newline=''))]
```
words for words => `line for line`

paulghaddad replied:
`new_words` => `new_lines`

paulghaddad reviewed (commented):
https://docs.python.org/3/library/csv.html

amymhaddad committed:
Update v2_ex46.py

amymhaddad commented on v2_ex46.py:
```diff
@@ -0,0 +1,27 @@
+for word in unique_words:
+    frequency = string_of_words.count(word)
+    final_output = f"{word}: {frequency}\n"
+    print(final_output.strip())
```
@paulghaddad The output prints out as it should. However, the most frequently used word should appear first (ie, badger: 7
mushroom: 2
...

I'm having trouble sorting this data so the most frequently used word appears first. thoughts?

Also, how can I replace the number of repeats with aestricks:
Ex: 2 ---> **

It seems like I need to use enumerate to index through each number. But I'm not having success with this. Is enumerate the right way to go?

paulghaddad replied:
So your `set` data structure won't allow you to do this. I'd recommend creating an _ordered_ dictionary of `word: frequency` pairs. Using an ordered dict will allow you to then loop through them from most frequent to least frequent. You can do so by iterating over `list_of_words`. I'd recommend this as practice. 

Once this is working, I'd recommend learning a built-in python data structure: [Counter](https://docs.python.org/3/library/collections.html#collections.Counter). This will under the hood create a similar data structure as your ordered dict, but give you some handy methods, such as [most_common](https://docs.python.org/3/library/collections.html#collections.Counter.most_common). This is exactly what you want as it will iterate from most to least common.

So I'd recommend one commit using the first method and a second commit for the second method.

amymhaddad committed:
Update v2_ex46.py

amymhaddad commented on v2_ex46.py:
```diff
@@ -0,0 +1,33 @@
+
+#Iteration 1:
+import csv
+from collections import OrderedDict
```
@paulghaddad Pylint says: Unused OrderedDict imported from collections. Why does it say this?

paulghaddad replied:
Not sure to be honest...maybe you don't need to import it? Maybe it's confused...

amymhaddad committed:
Update v2_ex46.py

amymhaddad committed:
Update v2_ex46.py

amymhaddad commented:
@paulghaddad Both iterations are included in this file!

paulghaddad commented on v2_ex46.py:
```diff
@@ -0,0 +1,59 @@
+
+    frequency_tracker = {}
+    for word in unique_words:
+        if word in list_of_words:
```
I'm actually not sure if you need line 26. Since you created `unique_words` from `list_of_words`, you know for sure that `word` is in the `list_of_words`. Second, I think you can just iterate through the `unique_words` set and then count each word in the `list_of_words`:

```
frequency_tracker = {}
for word in unique_words:
  frequency = list_of_words.count(word)
  frequency_tracker[word] = frequency
```

That eliminates some unneeded work. I like your approach of using `count` instead of iterating through `list_of_words` and incrementing a dictionary count `+= 1` each time you encounter a word (which is the approach most people take in this situation).

paulghaddad commented on v2_ex46.py:
```diff
@@ -0,0 +1,59 @@
+            frequency = list_of_words.count(word)
+            frequency_tracker[word] = frequency
+
+    frequencies = OrderedDict(sorted(frequency_tracker.items(), key=lambda t: t[1], reverse=True))
```
excellent

paulghaddad commented on v2_ex46.py:
```diff
@@ -0,0 +1,59 @@
+
+    for word, count in frequencies.items():
+        astericks = '*' * count
+        print(f"{word}: {astericks}")
```
👏

paulghaddad commented on v2_ex46.py:
```diff
@@ -0,0 +1,59 @@
+
+    counters = Counter(list_of_words).most_common()
+
+    frequently_used_words_with_counter = [f"{word}: {count}" for word, count in counters]
```
It doesn't appear that you're using this variable anywhere.

paulghaddad replied:
Good list comprehension though

paulghaddad commented on v2_ex46.py:
```diff
@@ -0,0 +1,59 @@
+            string_of_words += row + ' '
+            list_of_words = string_of_words.split()
+
+    counters = Counter(list_of_words).most_common()
```
I think a better name is `word_frequency_counter`. `counters` implies more than one counter (and you only have one), and it's pretty generic: it doesn't tell you _what_ you're counting.

paulghaddad reviewed (changes requested):
Nice work here...just a few deletions modifications and you should be good to go

amymhaddad committed:
Update v2_ex46.py

amymhaddad committed:
Update v2_ex46.py

amymhaddad commented:
@paulghaddad Good feedback! I removed the unused code that was on line 26 and the list comprehension that was on line 53. I also updated the variable name at 51.

paulghaddad reviewed (approved):
Really nice job on this! Proud of you! 💯
